torch/nn/init.py

```diff
@@ -525,3 +527,104 @@ def deprecated_init(*args, **kwargs):
+                if not ((str(char) >= '0' and str(char) <= '9') or str(char) == '.'):
+                    raise ValueError("Invalid version, must be a version string (e.g. '1.7.0') or tuple of integers")
+        v = version.split('.')
+        version = (int(v[0]), int(v[1]), int(v[2][0]))
+    else:
+        raise TypeError("Invalid version, must be a version string (e.g. '1.7.0') or tuple of integers")
+
+    # version, should be less than torch.__version__
+    if version > _torch_version:
+        raise ValueError(f"version {version} should be less than torch version {_torch_version}")
+    return version
+
+
+@contextlib.contextmanager
+def init_version(version=None, use_master=False):
+    r"""Context manager to use a specific version of initialization for `nn.modules`.
+    By default, the pytorch initialization used till version 1.6.1 is used.
+
+    Args:
+        version : Union[Tuple[int, int, int], str], which pytorch version to use for initializing nn.modules.
+            The format should bea version string (major,minor,micro) or a tuple of integers. Examples of valid
+            version are (1,7,0), '1.7.1'.
+        use_master : bool, If True, then the latest initialization scheme is used ignoring the value of `version`
+            (default=False)
+
+    Examples:
+        >>> with nn.init.init_version() as version:
+        >>>     # use the default pytorch initialization used till pytorch version 1.6.1
+        >>>     # This is the default behaviour.
+        >>>     model = nn.Sequential(...)
+        >>>
+        >>> with nn.init.init_version(use_master=True) as version:
+        >>>     # use latest initialization
+        >>>     model = nn.Sequential(...)
+        >>>
+        >>> with nn.init.init_version(version='1.7.1') as version:
+        >>>     model = nn.Sequential(...)
+
+    Note:
+        For researchers, who want to make their models available to the community, avoid hard-coding version
+        number as `with init_version('my.version.number')` in your code. The reason being when a new
+        initialization is released in pytorch, people would have to change the source code of your repo
+        in order to use that initialization. Instead, log/print a warning, to tell the community that your
+        model was built using this version of initialization scheme.
+
+        This can be illustrated as follows:
+        1. You release a model with version='1.7.0' hard-coded in your source code.
+            ```python
+            with nn.init.init_version('1.7.0') as version:
+                model = MyModel(...)
+            ```
+        2. Now in PyTorch=1.42.0 release a new initialization scheme is introduced.
+        3. To use this new initialization scheme, people would have to change the source code of your repo.
+        This can be a little inconvinient. 
+        4. To avoid this situation you can log a warning during model initialization that our model used this 
```

Comment:
the warning seems overly prescriptive.  What do you think researchers are concerned about where they would want to warn?

Reply:
It was suggested by @elistevens 

> The other way around, actually. If I write a model and hardcode 1.7.0 inits into it, then when 2.3.0 comes out with a better init scheme, someone who wants to try applying that new init scheme to my old model has to change the source code.

Reply:
I think the warning is very much wrong as it is currently written, too.

- People wanting to keep the init fixed regardless of PyTorch updates should use a fixed init version.
- Otherwise, one should not.

If anything, I would expect researchers releasing their code to be of the first sort, to facilitate reproducing their results. (To be honest, I would expect most researchers to currently init the things themselves, but that is something else.)
But I also agree with Greg that one might just leave people to choose what they want regardless of the job description.

Also, I would expect that research code basically sets the init somewhere globally and not as much use a context manager.

Reply:
Removed the warning.

Renamed various things:
* `_init_version` variable -> `init_version`
* `_parse_version` function -> `parse_version`
* `init_version` context manager -> `use_init_version`

> Also, I would expect that research code basically sets the init somewhere globally and not as much use a context manager.

Now, we can do `init.init_version = (1,4,1)`. This just assumes the user passes the input as `Tuple[int, int, int]`. I think we don't have to worry about wrong input here. Added this in the example of `use_init_version`.